### PR TITLE
Upgrade Payum up to 0.14

### PR DESCRIPTION
**Pull request by makasim**

### CODE REVIEW

**Comment by antonioperic** [comment]
great feature :+1:

File: src/Sylius/Bundle/PayumBundle/Model/PaymentConfig.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+*
+* For the full copyright and license information, please view the LICENSE
+* file that was distributed with this source code.
+*/
+
+namespace Sylius\Bundle\PayumBundle\Model;
+
+use Payum\Core\Model\PaymentConfigInterface;
+
+class PaymentConfig implements PaymentConfigInterface
```
Comment (makasim) [inline src/Sylius/Bundle/PayumBundle/Model/PaymentConfig.php]: @Sylius/core-team I am almost done with ability to configure payments gateways in the backend. The last thing is to provide a backend part for this model `PaymentConfig`. The form type must be this one `payum_payment_config ` see: https://github.com/Payum/PayumBundle/blob/master/Resources/config/form.xml#L16

Comment (winzou) [inline src/Sylius/Bundle/PayumBundle/Model/PaymentConfig.php]: The dynamic payment part should be added in another PR.
If everything is good as is, it's good to merge (just requires rebase).

**Comment by pjedrzejewski** [comment]
Cool! The PaymentConfig is persisted in the database? As an entity or serialized or how do you want to store it?

**Comment by pjedrzejewski** [comment]
Ok forget my question, now I see. I will see what I can figure out.

**Comment by makasim** [comment]
PaymentConfig could be somehow merged with Sylius PaymentMethod but I failed to do that.

**Comment by winzou** [comment]
ping @makasim :)

**Comment by makasim** [comment]
> Cool! The PaymentConfig is persisted in the database? As an entity or serialized or how do you want to store it?

It is an entity. heres the schema: https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/Doctrine/Resources/mapping/PaymentConfig.orm.xml

**Comment by makasim** [comment]
rebased and now require stable 0.14 version of Payum. 

I have not had chance to test it, hope to do it today in the evening.

**Comment by makasim** [comment]
Cannot test this PR because of an error while running  `sylius:install` 

https://www.dropbox.com/s/hwbftscc3lz2b34/Screenshot%202015-02-25%2017.02.13.png?dl=0

**Comment by pjedrzejewski** [comment]
@makasim Should be fixed in master, could you check again? https://github.com/Sylius/Sylius/commit/ffa93103627d0d054a04063c6285ff96d9224c9d

**Comment by makasim** [comment]
@pjedrzejewski I have ffa9310 locally but it does not help, for now commented out the code line .

**Comment by makasim** [comment]
If I run the command `./app/console sylius:install` (database already exist) everything works fine. but if I drop table before it fails the fatal I posted above.

**Comment by makasim** [comment]
Checked:

* [x] PaypalExpressCheckout
* [ ] Be2bill Offsite 
* [ ] Be2Bill CreditCard
* [ ] Dummy

**Comment by makasim** [comment]
@Sylius/core-team almost done. One thing left.

Everybody is very welcome to review it again.

File: src/Sylius/Bundle/PayumBundle/Resources/config/driver/doctrine/orm.xml [hunk 2]
```diff
@@ -34,6 +35,20 @@
...
+            <argument type="service">
+                <service
+                    factory-service="sylius.manager.payment_config"
+                    factory-method="getClassMetadata"
+                    class="Doctrine\ORM\Mapping\ClassMetadata"
+                >
+                    <argument>%sylius.model.payment_config.class%</argument>
+                </service>
+            </argument>
+        </service>
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/PayumBundle/Resources/config/driver/doctrine/orm.xml]: This is not needed, will be generated automatically.

Comment (makasim) [inline src/Sylius/Bundle/PayumBundle/Resources/config/driver/doctrine/orm.xml]: @pjedrzejewski  both? manager and repository?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/PayumBundle/Resources/config/driver/doctrine/orm.xml]: Both! Parameter as well.

Comment (makasim) [inline src/Sylius/Bundle/PayumBundle/Resources/config/driver/doctrine/orm.xml]: @pjedrzejewski done

**Comment by makasim** [comment]
@pjedrzejewski tested everything manually and fixed phpspecs.

**Comment by makasim** [comment]
@pjedrzejewski ping

File: src/Sylius/Bundle/PayumBundle/Payum/Be2bill/Action/NotifyAction.php [hunk 3]
```diff
@@ -66,6 +66,18 @@ public function __construct(
 
     /**
      * {@inheritDoc}
+     */
+    public function setApi($api)
+    {
+        if (false == $api instanceof Api) {
```
Comment (stloyd) [inline src/Sylius/Bundle/PayumBundle/Payum/Be2bill/Action/NotifyAction.php]: Just:
```php
if (!$api instanceof Api) {}

**Comment by stloyd** [comment]
Just one comment, after that please squash those commits a bit, thanks!

:+1:

**Comment by makasim** [comment]
\-:

**Comment by makasim** [comment]
sigh

**Comment by makasim** [comment]
@pjedrzejewski @stloyd done

**Comment by makasim** [comment]
@pjedrzejewski ping

**Comment by pjedrzejewski** [comment]
Just one quick question, couldn't we save this configuration directly on PaymentMethod, why the need for separate PaymentConfig?

**Comment by makasim** [comment]
@pjedrzejewski it could be but in that case we would not able to reuse DynamicRegistry and PaymentConfigType form type. Ofcource similar logic could be implemented for PaymentMethod too but not a good approach imo.


better to implement PaymentConfigInterface on top of PaymentMethod so Payum works with it as PaymentConfig and Sylius works with it as PaymentMethod. I tried to do that but failed.

**Comment by makasim** [comment]
@pjedrzejewski ping

**Comment by makasim** [comment]
@pjedrzejewski ping

**Comment by pjedrzejewski** [comment]
Awesome work Maksim, thanks! Now next step is handling the PaymentConfig in forms (backend)? I could help with that, unless you have some idea already!

**Comment by makasim** [comment]
There is a form type https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/Symfony/Form/Type/PaymentConfigType.php I use in sonata admin extension https://github.com/Payum/PayumBundle/blob/master/Sonata/PaymentConfigAdmin.php

**Comment by winzou** [comment]
Great update thanks @makasim 

I have 2 questions regarding your implementation:
  1. Is possible to access the PaymentConfig $config attribute from a CaptureAction?
  2. Is it possible to define this $config from the yaml configuration, and how?

Thanks

**Comment by makasim** [comment]
> Is it possible to access the PaymentConfig $config attribute from a CaptureAction, and how?

Nope it used only by a [DynamicRegistry](https://github.com/Payum/Payum/blob/master/src/Payum/Core/Registry/DynamicRegistry.php#L48) to create a payment instance. What do you want to get from the config in capture action?

> Is it possible to define this $config from the yaml configuration, and how?

Not sure I get the question. In theory yes but you have to take care of that and pass it to a payment factory.
